### ISSUE TITLE
Feature Request: browse Kmacro ring (with example)

### ISSUE THREAD

**Issue by okamsn** [abo-abo/swiper#2173]
Hello,

I feel like this would be useful, because it can be difficult to remember where a macro is in the ring. Below is an example of what I have been using. It can be improved, because I am not so good with Lisp.

```lisp
(defun counsel-kmacro ()
    "Interactively choose a keyboard macro to run."
    (interactive)
    (ivy-read "Keyboard macro: "
              (cons  (propertize (concat "Last Macro: " (format-kbd-macro last-kbd-macro 1)) 'actual-kmacro last-kbd-macro)
                     (mapcar (lambda (kmacro)
                               ;; Add the actual kmacro as a property of the displayed string representing the macro.
                               (let ((actual-macro (nth 0 kmacro)))
                                 (propertize (concat "---------------\n" (format-kbd-macro actual-macro 1)) 'actual-kmacro actual-macro)))
                             kmacro-ring)))
              :require-match t
              :sort nil
              :action (lambda (formatted-kmacro)
                        (kmacro-call-macro 1 t nil (nth 1 (text-properties-at 0 formatted-kmacro))))
              :caller 'counsel-kmacro)

 (defun counsel-kmacro-delete-action (formatted-kmacro)
   "Delete a keyboard macro from `kmacro-ring', or delete the last macro."
   (let ((actual-macro (text-properties-at 0 formatted-kmacro)))
     (if (eq actual-macro last-kbd-macro)
         (setq last-kbd-macro nil)
       (setq kmacro-ring (remove actual-macro kmacro-ring)))))

 (ivy-set-actions #'counsel-kmacro '(("d" counsel-kmacro-delete-action "delete")))
```
There is delete action because of saving macro ring using `savehist`.

Thank you.

**Comment by abo-abo** [comment]
Hi. Nice idea, and your code looks good. You probably missed a paren when copying it here :). Please make a PR.
If you put just `counsel-macro` in the PR, I can merge it right away, since it's below 15 lines. You'll need an Emacs CA for the rest, see CONTRIBUTING.org for more info.

**Comment by okamsn** [comment]
Hello,

I have signed the Emacs CA paperwork, and am still improving this idea.

What is the proper way to update the displayed list after an alternative action from `C-M-o` has been taken?

Thank you.

**Comment by abo-abo** [comment]
Please see how `ivy-switch-buffer-kill` handles it, maybe just calling `ivy--kill-current-candidate` is enough.

**Comment by abo-abo** [comment]
Fixed in #2280. Closing. Thanks.
